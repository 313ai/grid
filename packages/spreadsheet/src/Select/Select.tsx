import React from "react";
import { IconButton, Tooltip } from "./../styled";
import Downshift from "downshift";
import { Box, useTheme, useColorMode } from "@chakra-ui/core";
import { DARK_MODE_COLOR, DARK_MODE_COLOR_LIGHT } from "../constants";
import { KeyCodes } from "@rowsncolumns/grid/dist/types";

export interface SelectProps {
  showInput?: boolean;
  options: Option[];
  value?: Option;
  onChange?: (value: Option | null) => void;
  format?: (value: string) => any;
  inputWidth?: number;
  enableInput?: boolean;
}
export interface Option {
  value: string | number;
  label: string | number;
}

const Select: React.FC<SelectProps> = props => {
  const {
    options,
    value,
    onChange,
    format,
    inputWidth = 44,
    enableInput = true
  } = props;
  const theme = useTheme();
  const { colorMode } = useColorMode();
  const isLight = colorMode === "light";
  const inputBorderColor = isLight
    ? theme.colors.gray[300]
    : theme.colors.gray[600];
  const inputBgColor = isLight ? theme.colors.white : DARK_MODE_COLOR;
  const inputColor = isLight ? DARK_MODE_COLOR : theme.colors.white;
  const dropdownBgColor = isLight ? theme.colors.white : theme.colors.gray[700];
  return (
    <Downshift
      selectedItem={value}
      onChange={sel => onChange?.(sel)}
      itemToString={item => (item ? item.value.toString() : "")}
    >
      {({
        getInputProps,
        getMenuProps,
        getRootProps,
        isOpen,
        getItemProps,
        inputValue,
        highlightedIndex,
        selectedItem,
        selectItem,
        closeMenu,
        getToggleButtonProps,
        openMenu
      }) => {
        const inputProps = getInputProps();
        return (
          <Box position="relative" {...getRootProps()}>
            <Box display="flex" alignItems="center">
              {enableInput ? (
                <input
                  style={{
                    width: inputWidth,
                    height: 24,
                    paddingLeft: 4,
                    paddingRight: 2,
                    borderColor: inputBorderColor,
                    background: inputBgColor,
                    color: inputColor,
                    borderStyle: "solid",
                    borderWidth: 1,
                    fontSize: 12
                  }}
                  {...inputProps}
                  onKeyDown={(e: React.KeyboardEvent<HTMLInputElement>) => {
                    if (inputProps && inputProps.onKeyDown)
                      inputProps.onKeyDown(e);
                    if (
                      e.which === KeyCodes.Enter &&
                      highlightedIndex === null
                    ) {
                      const value = format?.(e.currentTarget.value);
                      selectItem?.({ value, label: value });
                      closeMenu();
                    }
                  }}
                  onFocus={() => openMenu()}
                />
              ) : (
                <Tooltip
                  hasArrow
                  aria-label={inputValue || ""}
                  label={inputValue || ""}
                  placement="bottom-start"
                >
                  <Box
                    cursor="pointer"
                    pl={2}
                    pr={2}
                    title={inputValue}
                    {...getToggleButtonProps()}
                    whiteSpace="nowrap"
                    overflow="hidden"
                    textOverflow="ellipsis"
                    fontSize={12}
                    width={inputWidth}
                  >
                    {inputValue}
                  </Box>
                </Tooltip>
              )}
              <IconButton
                color={inputColor}
                height={5}
                minWidth={5}
                aria-label="Open"
                icon="chevron-down"
                size="sm"
                fontSize={16}
                {...getToggleButtonProps()}
              />
            </Box>
            <Box
              background={dropdownBgColor}
              shadow="md"
              position="absolute"
              top="100%"
              zIndex={1}
              borderRadius={5}
              {...getMenuProps()}
              pb={1}
              pt={1}
            >
              {isOpen &&
                options.map((item, index) => {
                  const { label, value } = item;
                  return (
                    <Box
                      fontSize={12}
                      padding={2}
                      pl={3}
                      pr={3}
                      whiteSpace="nowrap"
                      {...getItemProps({
                        key: value,
                        index,
                        item,
                        style: {
                          cursor: "pointer",
                          backgroundColor:
                            highlightedIndex === index
                              ? isLight
                                ? theme.colors.gray[100]
                                : "rgba(255,255,255,0.06)"
                              : dropdownBgColor,
                          fontWeight: selectedItem === item ? "bold" : "normal"
                        }
                      })}
                    >
                      {label}
                    </Box>
                  );
                })}
            </Box>
          </Box>
        );
      }}
    </Downshift>
  );
};

export default Select;
