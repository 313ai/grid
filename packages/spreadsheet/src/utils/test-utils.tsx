import React from "react";
import { render } from "@testing-library/react";
import { Stage, Layer } from "react-konva";
import { ThemeProvider, theme } from "@chakra-ui/core";
import Konva from "konva";

const KonvaWrapper = ({ children }) => {
  return (
    <Stage>
      <ThemeWrapper>
        <Layer>{children}</Layer>
      </ThemeWrapper>
    </Stage>
  );
};

const konvaRenderer = (ui) => render(ui, { wrapper: KonvaWrapper });
const domRenderer = (ui) => render(ui, { wrapper: ThemeWrapper });

export * from "@testing-library/react";
export { konvaRenderer, domRenderer };

export const ThemeWrapper = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};
