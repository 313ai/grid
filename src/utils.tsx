import React from "react";
import { Line, Rect } from "react-konva";
import { ShapeConfig } from "konva/types/Shape";

interface BoxProps extends ShapeConfig {
  onCrosshairMouseDown?: (e: React.MouseEvent<HTMLDivElement>) => void;
  onCrosshairMouseMove?: (e: React.MouseEvent<HTMLDivElement>) => void;
  onCrosshairMouseUp?: (e: React.MouseEvent<HTMLDivElement>) => void;
}

/**
 * Create a box with custom top/right/bottom/left colors and widths
 * @param param0
 */
export const createBox = ({
  x = 0,
  y = 0,
  width = 0,
  height = 0,
  fill,
  stroke,
  strokeLeftColor = stroke,
  strokeTopColor = stroke,
  strokeRightColor = stroke,
  strokeBottomColor = stroke,
  strokeWidth = 0,
  strokeTopWidth = strokeWidth,
  strokeRightWidth = strokeWidth,
  strokeBottomWidth = strokeWidth,
  strokeLeftWidth = strokeWidth,
  key,
}: BoxProps) => {
  const commonProps = {
    perfectDrawEnabled: false,
    shadowForStrokeEnabled: false,
    hitStrokeWidth: 0,
    listening: false,
    lineCap: "square",
  };
  const lines = [
    <Line
      points={[x, y, x + width, y]}
      stroke={strokeTopColor}
      strokeWidth={strokeTopWidth}
      key="top"
      {...commonProps}
    />,
    <Line
      points={[x + width, y, x + width, y + height]}
      stroke={strokeRightColor}
      strokeWidth={strokeRightWidth}
      key="right"
      {...commonProps}
    />,
    <Line
      points={[x + width, y + height, x, y + height]}
      stroke={strokeBottomColor}
      strokeWidth={strokeBottomWidth}
      key="bottom"
      {...commonProps}
    />,
    <Line
      points={[x, y + height, x, y]}
      stroke={strokeLeftColor}
      strokeWidth={strokeLeftWidth}
      key="left"
      {...commonProps}
    />,
  ];

  return (
    <React.Fragment key={key}>
      {lines}
      {fill && (
        <Rect
          fill={fill}
          x={x}
          y={y}
          width={width}
          height={height}
          {...commonProps}
        />
      )}
    </React.Fragment>
  );
};

export const FillHandle: React.FC<BoxProps> = ({
  x = 0,
  y = 0,
  stroke,
  onMouseDown,
}) => {
  if (x === 0 || y === 0) return null;
  const crossHairSize = 8;
  const crossHairBorderSize = 1;
  return (
    <div
      style={{
        position: "absolute",
        left: x - crossHairSize / 2,
        top: y - crossHairSize / 2,
        width: crossHairSize,
        height: crossHairSize,
        border: `${crossHairBorderSize}px white solid`,
        background: stroke,
        cursor: "crosshair",
      }}
      onMouseDown={onMouseDown}
    />
  );
};

export const createHTMLBox = ({
  x = 0,
  y = 0,
  width = 0,
  height = 0,
  fill,
  stroke,
  strokeLeftColor = stroke,
  strokeTopColor = stroke,
  strokeRightColor = stroke,
  strokeBottomColor = stroke,
  strokeWidth = 0,
  strokeTopWidth = strokeWidth,
  strokeRightWidth = strokeWidth,
  strokeBottomWidth = strokeWidth,
  strokeLeftWidth = strokeWidth,
  key,
  strokeStyle = "solid",
}: BoxProps) => {
  const commonProps = {};
  width = width - Math.floor(strokeWidth);
  y = y - Math.ceil(strokeWidth / 2);
  const lines = [
    <div
      style={{
        position: "absolute",
        left: x,
        top: y,
        width: width,
        height: strokeTopWidth,
        borderWidth: 0,
        borderColor: strokeTopColor,
        borderTopWidth: strokeTopWidth,
        borderStyle: strokeStyle,
      }}
      key="top"
      {...commonProps}
    />,
    <div
      style={{
        position: "absolute",
        left: x + width,
        top: y,
        width: strokeRightWidth,
        height: height,
        borderWidth: 0,
        borderColor: strokeRightColor,
        borderRightWidth: strokeRightWidth,
        borderStyle: strokeStyle,
      }}
      key="right"
      {...commonProps}
    />,
    ,
    <div
      style={{
        position: "absolute",
        left: x,
        top: y + height,
        width: width + strokeTopWidth,
        height: strokeBottomWidth,
        borderWidth: 0,
        borderColor: strokeBottomColor,
        borderBottomWidth: strokeBottomWidth,
        borderStyle: strokeStyle,
      }}
      key="bottom"
      {...commonProps}
    />,
    <div
      style={{
        position: "absolute",
        left: x,
        top: y,
        width: strokeLeftWidth,
        height: height,
        borderWidth: 0,
        borderColor: strokeLeftColor,
        borderLeftWidth: strokeLeftWidth,
        borderStyle: strokeStyle,
      }}
      key="left"
      {...commonProps}
    />,
  ];

  return (
    <React.Fragment key={key}>
      {lines}
      {fill && (
        <div
          style={{
            position: "absolute",
            top: y,
            left: x,
            height,
            width,
            backgroundColor: fill,
            userSelect: "none",
            pointerEvents: "none",
          }}
          {...commonProps}
        />
      )}
    </React.Fragment>
  );
};
