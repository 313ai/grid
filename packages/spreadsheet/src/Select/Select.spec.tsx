import React, { useState, useRef } from "react";
import { domRenderer, cleanup, fireEvent, act } from "./../utils/test-utils";
import Select from "./Select";

describe("Select", () => {
  afterEach(cleanup);
  const options = [
    {
      label: 'A',
      value: 0
    },
    {
      label: 'B',
      value: 1
    }
  ]
  it("renders select dropdown", () => {
    const renderGrid = () => domRenderer(<Select options={options} />);
    expect(renderGrid).not.toThrow();
  });
  it("matches snapshot", () => {
    const { asFragment } = domRenderer(<Select options={options} />);
    expect(asFragment()).toMatchSnapshot();
  });

  // it("can trigger onchange", () => {
  //   const onChange = jest.fn();
  //   const { getByLabelText } = domRenderer(<FormulaBar onChange={onChange} />);
  //   const input = getByLabelText("value-input") as HTMLInputElement;
  //   fireEvent.change(input, { target: { value: "hello" } });
  //   expect(onChange).toBeCalled();
  //   expect(onChange).toBeCalledWith("hello");
  // });
});
