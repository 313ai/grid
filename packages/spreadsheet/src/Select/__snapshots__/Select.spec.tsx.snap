// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-79elbk"
  >
    <div
      class="css-70qvj9"
    >
      <input
        style="width: 44px; height: 24px; padding-left: 4px; padding-right: 2px; border-color: #cbd5e0; background: rgb(255, 255, 255); color: rgb(26, 32, 44); border-style: solid; border-width: 1px; font-size: 12px;"
        value=""
      />
      <button
        aria-label="Open"
        class="css-qvndex-IconButton"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="css-1im46kq"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      class="css-b0ps0v"
    />
  </div>
</DocumentFragment>
`;
